cdap-gateway/src/main/java/co/cask/cdap/gateway/router/RouterPathLookup.java

```diff
@@ -218,24 +220,26 @@ private boolean matches(String[] actual, String... expected) {
   }
 
   /**
-   * Determines if the actual ends with expected. If expected contains '?' at the end the the last element of actual
-   * is ignored as ? is considered to match one element regardless of what it is
+   * Determines if the actual ends with expected. '?' in expected is considered to match
+   * one element regardless of what it is.
    */
   @VisibleForTesting
   boolean endsWith(String[] actual, String... expectedEnd) {
     if (expectedEnd.length > actual.length) {
       return false;
     }
+
     int offset = 1;
-    if (expectedEnd[expectedEnd.length - offset].equals("?")) {
-      offset++;
-    }
-    while (expectedEnd.length - offset >= 0) {
-      if (expectedEnd[expectedEnd.length - offset].equalsIgnoreCase(actual[actual.length - offset])) {
+    while (offset <= expectedEnd.length) {
+      if (expectedEnd[expectedEnd.length - offset].equals("?")) {
```

Comment:
This can be further simplified by combining the two `if` conditions into one, since the `if` body is the same in both cases - 
```
 if (expectedEnd[expectedEnd.length - offset].equals("?") 
     || expectedEnd[expectedEnd.length - offset].equalsIgnoreCase(actual[actual.length - offset])) {
    offset++;
} else {
    return false;
}

return true;
```

Reply:
fixed.